[FIX] stock: don't rollback scheduler if it succeeeded

luisg123v opened the pull request:
Currently, when the stock scheduler is run, a new cursor is created and
the result is rolled back even if all tasks succeeded. Even though tasks
themselves are not rolled back because they are run using new other
cursors, rolling it back is not correct and it makes harder to test the
scheduler feature.

The rollback was introduced in da6bd90e by mistake.

This commit makes the transaction to be rolled back only when the
current transaction actually failed.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/175621.png)](https://mergebot.odoo.com/odoo/odoo/pull/175621)

luisg123v commented:
Hi @moylop260,

What do you think about this change?

CC @desdelinux @teposteaj

moylop260 commented:
@luisg123v 

The reference to commit in the description is now linked correctly

I saw this rollback from old stable Odoo version 15.0:
https://github.com/odoo/odoo/blob/def96a8ee7f6430ce2b6c173544f3f1db8b602c0/addons/stock/wizard/stock_scheduler_compute.py#L40

Should we fix it from this version in order to be auto-FW using the bot

Or Should be fix it only for master version?

moylop260 commented:
@KangOl was this `rollback` added by mistake?

KangOl commented:
Yes, it was.

moylop260 commented:
@KangOl 

Thank you for reply

What is the best branch to make this fix?

and what person should we tag it?

ticodoo commented:
Hello @luisg123v @moylop260 this fix is acceptable for stable, but can you retarget it for v15? It is true that support for it will be dropped, but not for a couple more months and it should be a hopefully easy fw-port for the other versions. You can let me know when the PR is ready and I will merge it

luisg123v commented:
Hi @ticodoo,

Thanks for the advice. I've changed target branch to 15.0. I also noticed a clean forward-port is possible.

Could you merge this, please?

Regards,

ticodoo commented:
Thanks!

@robodoo r+

robodoo commented:
@luisg123v @ticodoo 'ci/runbot' failed on this reviewed PR.

ticodoo commented:
@luisg123v can you try pulling the latest v15 and rebasing again?

luisg123v commented:
@ticodoo done, rebased.  I had to modify one test because it relied on the effects of the old rollback.

Could you review again and comment if the modification is an acceptable solution, please?

luisg123v commented:
Hi @ticodoo, could you review/merge, please?

luisg123v commented:
Hi @ticodoo, do you agree with these changes? Any suggestions for this to be mergeable?

luisg123v commented:
Hi @KangOl, @ticodoo,

I did the following:
- Changed target branch to 16.0, as 15.0 is being deprecated.
- Removed changes to the module `test_main_flows` because they're not required anymore in 16.0.
Could you review, please?

Regards,

ticodoo commented:
Whoops, thanks! Didn't see that the test no longer fails in v16 (I was planning on having to spend a lot of time figuring out why it was failing).

Note that the wizard was removed in 18.0 (in this PR https://github.com/odoo/odoo/pull/179005), so this fix isn't needed beyond that

@robodoo r+ up to saas-17.4

robodoo commented:
Forward-porting to 'saas-17.4'.

luisg123v commented:
@ticodoo thanks for the info and for the merge!

Just for the record, the test was failing because, at some point after the scheduler was run, it re-reserved quantities, which was not necessary after this fix because reservation done by the scheduled was not being rolled back anymore.
